types/priv_validator/socket.go

```diff
@@ -195,166 +240,206 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
+// RemoteSigner.
+func NewRemoteSigner(
 	logger log.Logger,
 	chainID, socketAddr string,
-	maxConnections int,
 	privVal PrivValidator,
 	privKey *crypto.PrivKeyEd25519,
-) *PrivValidatorSocketServer {
-	proto, addr := cmn.ProtocolAndAddress(socketAddr)
-	pvss := &PrivValidatorSocketServer{
-		proto:          proto,
-		addr:           addr,
-		maxConnections: maxConnections,
-		privKey:        privKey,
-		privVal:        privVal,
-		chainID:        chainID,
+) *RemoteSigner {
+	rs := &RemoteSigner{
+		addr:         socketAddr,
+		chainID:      chainID,
+		connDeadline: time.Second * defaultConnDeadlineSeconds,
+		connRetries:  defaultDialRetries,
+		privKey:      privKey,
+		privVal:      privVal,
 	}
-	pvss.BaseService = *cmn.NewBaseService(logger, "privValidatorSocketServer", pvss)
-	return pvss
+
+	rs.BaseService = *cmn.NewBaseService(logger, "RemoteSigner", rs)
+
+	return rs
 }
 
 // OnStart implements cmn.Service.
-func (pvss *PrivValidatorSocketServer) OnStart() error {
-	ln, err := net.Listen(pvss.proto, pvss.addr)
+func (rs *RemoteSigner) OnStart() error {
+	conn, err := rs.connect()
 	if err != nil {
+		rs.Logger.Error("OnStart", "err", errors.Wrap(err, "connect"))
+
 		return err
 	}
 
-	pvss.listener = netutil.LimitListener(ln, pvss.maxConnections)
-
-	go pvss.acceptConnections()
+	go rs.handleConnection(conn)
 
 	return nil
 }
 
 // OnStop implements cmn.Service.
-func (pvss *PrivValidatorSocketServer) OnStop() {
-	if pvss.listener == nil {
+func (rs *RemoteSigner) OnStop() {
+	if rs.conn == nil {
 		return
 	}
 
-	if err := pvss.listener.Close(); err != nil {
-		pvss.Logger.Error("OnStop", "err", errors.Wrap(err, "closing listener failed"))
+	if err := rs.conn.Close(); err != nil {
+		rs.Logger.Error("OnStop", "err", errors.Wrap(err, "closing listener failed"))
 	}
 }
 
-func (pvss *PrivValidatorSocketServer) acceptConnections() {
-	for {
-		conn, err := pvss.listener.Accept()
+func (rs *RemoteSigner) connect() (net.Conn, error) {
+	retries := defaultDialRetries
+
+RETRY_LOOP:
+	for retries > 0 {
+		// Don't sleep if it is the first retry.
+		if retries != defaultDialRetries {
+			time.Sleep(rs.connDeadline)
+		}
+
+		retries--
+
+		conn, err := cmn.Connect(rs.addr)
 		if err != nil {
-			if !pvss.IsRunning() {
-				return // Ignore error from listener closing.
-			}
-			pvss.Logger.Error(
-				"acceptConnections",
-				"err", errors.Wrap(err, "failed to accept connection"),
+			rs.Logger.Error(
+				"connect",
+				"addr", rs.addr,
+				"err", errors.Wrap(err, "connection failed"),
 			)
-			continue
+
+			continue RETRY_LOOP
 		}
 
 		if err := conn.SetDeadline(time.Now().Add(connDeadline)); err != nil {
-			pvss.Logger.Error(
-				"acceptConnetions",
+			rs.Logger.Error(
+				"connect",
 				"err", errors.Wrap(err, "setting connection timeout failed"),
 			)
 			continue
 		}
 
-		if pvss.privKey != nil {
-			conn, err = p2pconn.MakeSecretConnection(conn, pvss.privKey.Wrap())
+		if rs.privKey != nil {
```

Comment:
require it

Reply:
we also probably want to authenticate. ie. we should be dialing a `ID@host:port` and then after the secret connection is setup, check that the ID (ie. hash of the remote pubkey) of the connection matches the ID we dialed (see how p2p does it)